src/storage/src/source/mysql/replication.rs

```diff
@@ -199,29 +236,63 @@ pub(crate) fn render<G: Scope<Timestamp = TransactionId>>(
                         // sent within the heartbeat interval. This means that we can safely advance the
                         // frontier once since we know there were no more events for the last sent GTID.
                         // See: https://dev.mysql.com/doc/refman/8.0/en/replication-administration-status.html
-                        if advance_on_heartbeat {
-                            data_output.give_container(&data_cap_set[0], &mut container).await;
-
-                            new_upper = new_upper + TransactionId::new(1);
-                            trace!(%id, "heartbeat: timely-{worker_id} advancing frontier to {new_upper:?}");
-                            upper_cap_set.downgrade([&new_upper]);
-                            data_cap_set.downgrade([&new_upper]);
-                            rewinds.retain(|_, (_, req)| data_cap_set[0].time() <= &TransactionId::new(req.snapshot_gtid_set.first().expect("at least one gtid").latest_transaction_id()));
-                            advance_on_heartbeat = false;
+                        if let Some(new_gtid) = new_gtid_upper {
+                            let new_upper = GtidPartition::new_singleton(
+                                new_gtid.interval().singleton().unwrap().clone(),
+                                *new_gtid.timestamp() + TransactionId::Active(1),
+                            );
+
+                            trace!(%id, "heartbeat: timely-{worker_id} advancing partition frontier to {new_upper:?}");
+                            active_partition_caps.get_mut(new_gtid.interval().singleton().unwrap()).unwrap()
+                                .downgrade(&new_upper);
+                            data_cap_set.downgrade(&*future_partitions);
+                            upper_cap_set.downgrade(&*future_partitions);
+
+                            // Retain any rewinds whose snapshot still contains partitions with transaction-ids greater
+                            // than their corresponding frontiers
+                            rewinds.retain(|_, (_, partitions)|
+                                partitions.iter().any(|(uuid, rewind_part)|
+                                    active_partition_caps.get(uuid).is_some_and(|active_cap| active_cap.tx_id() <= &rewind_part.timestamp())
+                            ));
```

Comment:
Same here, each rewind request should be associated with a frontier and this check becomes:
```suggestion
// We retain all rewind requests 
rewinds.retain(|_, (_, upper)| PartialOrder::less_equal(current_upper, snapshot_upper))
```


```